mmpose/models/losses/regression_loss.py

```diff
@@ -6,6 +6,94 @@
+
+    Args:
+        use_target_weight (bool): Option to use weighted MSE loss.
+            Different joint types may have different target weights.
+        size_average (bool): Option to average the loss by the batch_size.
+        residual (bool): Option to add L1 loss and let the flow
+            learn the residual error distribution.
+        q_dis (string): Option for the identity Q(error) distribution,
+            Options: "Laplace" or "Gaussian"
+    """
+
+    def __init__(self,
+                 use_target_weight=False,
+                 size_average=True,
+                 residual=True,
+                 q_dis='Laplace'):
+        super(RLELoss, self).__init__()
+        self.size_average = size_average
+        self.use_target_weight = use_target_weight
+        self.residual = residual
+        self.q_dis = q_dis
+
+        self.flow = RealNVP()
+
+    def _apply(self, fn):
+        self.flow.prior.loc = fn(self.flow.prior.loc)
+        self.flow.prior.scale_tril = fn(self.flow.prior.scale_tril)
+        self.flow.prior._unbroadcasted_scale_tril =\
+            fn(self.flow.prior._unbroadcasted_scale_tril)
+        self.flow.prior.covariance_matrix = \
+            fn(self.flow.prior.covariance_matrix)
+        self.flow.prior.precision_matrix = \
+            fn(self.flow.prior.precision_matrix)
+        self.flow.mask = fn(self.flow.mask)
+        self.flow.s._apply(fn)
+        self.flow.t._apply(fn)
+
+    def forward(self, output, target, target_weight=None):
+        """Forward function.
+
+        Note:
+            - batch_size: N
+            - num_keypoints: K
+            - dimension of keypoints: D (D=2 or D=3)
+
+        Args:
+            output (torch.Tensor[N, K, D*2]): Output regression,
+                    including coords and sigmas.
+            target (torch.Tensor[N, K, D]): Target regression.
+            target_weight (torch.Tensor[N, K, D]):
+                Weights across different joint types.
+        """
+        sigma = output[:, :, 2:].sigmoid()
+        coord = output[:, :, :2]
+
+        error = (coord - target) / sigma
+        # (B, K, 2)
+        log_phi = self.flow.log_prob(error.reshape(-1, 2))
+        log_phi = log_phi.reshape(target.shape[0], target.shape[1], 1)
+        log_sigma = torch.log(sigma).reshape(target.shape[0], target.shape[1],
+                                             2)
+        nf_loss = log_sigma - log_phi
+
+        if self.residual:
+            assert self.q_dis in ['Laplace', 'Gaussian']
+            if self.q_dis == 'Laplace':
+                loss_q = torch.abs(error)
+            else:
+                loss_q = error**2
+            loss = nf_loss + loss_q
+        else:
+            loss = nf_loss
+
+        if self.use_target_weight:
+            assert target_weight is not None
+            loss *= target_weight
+
+        if self.size_average:
+            loss /= len(loss)
+
+        return loss.sum()
```

Comment:
@Indigo6 Thanks for the detailed [explanation](https://github.com/open-mmlab/mmpose/pull/1259#discussion_r844574413) of the loss implementation. I am clear about Q2-Q4. 

About Q1, as you explained, `self.flow.log_prob(error)` is $-logP(\bar\mu)$. However in [`RealNVP.log_prob()`](https://github.com/open-mmlab/mmpose/pull/1259/files#diff-711f9876f89558df68e9c40a354cd40316a556a3922b998b0f8f2233494d95daR57-R59) it seems that the return value is the summation of $logP(\bar\mu)$ and $log_det$. The second term, according to Eq.5 in the paper, should be $log\ \hat\sigma$, which leads to my previous question that `log_phi` seems to already contain the $log\ \hat\sigma$ term and why it's added again in [L77](https://github.com/open-mmlab/mmpose/pull/1259/files#diff-9674ca04b72898ff995646744e5d26d5a8b6346de33d2c564d5f45ed9467e6b6R77). Please correct me if I misunderstand the math somewhere.

The rest parts look all good to me and we will very the model training this week. And if you don't mind, we may probably make some minor modifications and directly push them to this PR. Feel free to leave your comments and suggestions on these changes.

Reply:
> @Indigo6 Thanks for the detailed [explanation](https://github.com/open-mmlab/mmpose/pull/1259#discussion_r844574413) of the loss implementation. I am clear about Q2-Q4.
> 
> About Q1, as you explained, `self.flow.log_prob(error)` is −logP(μ¯)-logP(\bar\mu). However in [`RealNVP.log_prob()`](https://github.com/open-mmlab/mmpose/pull/1259/files#diff-711f9876f89558df68e9c40a354cd40316a556a3922b998b0f8f2233494d95daR57-R59) it seems that the return value is the summation of logP(μ¯)logP(\bar\mu) and logdetlog_det. The second term, according to Eq.5 in the paper, should be log σ^log\ \hat\sigma, which leads to my previous question that `log_phi` seems to already contain the log σ^log\ \hat\sigma term and why it's added again in [L77](https://github.com/open-mmlab/mmpose/pull/1259/files#diff-9674ca04b72898ff995646744e5d26d5a8b6346de33d2c564d5f45ed9467e6b6R77). Please correct me if I misunderstand the math somewhere.
> 
> The rest parts look all good to me and we will very the model training this week. And if you don't mind, we may probably make some minor modifications and directly push them to this PR. Feel free to leave your comments and suggestions on these changes.

@ly015 Hi, about Q1, there are two functions $f$ and $\phi$, and two corresponding $log|det(J(f^{-1}))|$ and $log|det(J(\phi^{-1}))|$. The first function stands for the reparameterization: 
$$
\mu = f(\bar\mu) = \bar\mu*\hat\sigma+\hat\mu
$$ where its $log|det(J(f^{-1}))|$ is $log \hat\sigma$.\\
The second function is the RealNVP:
$$
\bar\mu=\phi(z)
$$ where its $log|det(J(\phi^{-1}))|$ is not $log \hat\sigma$. In other words, `self.flow.log_prob(error)` do not include $log \hat\sigma$. \\

About model performance, it may be a little lower than the original implementation. After I simplified the implementation of log(Q) in commit [facb0d0](https://github.com/open-mmlab/mmpose/pull/1259/commits/facb0d01388b3216668ba3edb4bd3346c50c0bc6), I retrained the model and the performance degraded a bit. 

About direct modifications, it's ok for me.

Reply:
Thanks for your reply. About the performance degradation, do you think it's caused by the removal of `loss_sigma`?

Reply:
Yes, the removal of `loss_sigma` should be the reason, but I don't know why it caused performance degradation